Expose well-known certificate subject fields in Lua filter

zhishi opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: expose well-known certificate subject fields in Lua filter for https://github.com/envoyproxy/envoy/issues/35524
Additional Description:
Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

zhishi commented:
@ggreenway this is the draft PR.

ggreenway commented:
@zhishi thanks for posting this. Yes, I think this looks like a good way to model the the lua API for this; this all looks reasonable.

zhishi commented:
@ggreenway then I'll start working on the unit test and change it to a formal PR. thanks for reviewing!

zhishi commented:
@ggreenway  I have added unit test and the PR is good for review. I still have question about how to update document for the added interface, will that be automatically generated from comments or I should update some doc file in somewhere?

ggreenway reviewed (changes requested):
/wait

ggreenway commented:
> @ggreenway I have added unit test and the PR is good for review. I still have question about how to update document for the added interface, will that be automatically generated from comments or I should update some doc file in somewhere?

Docs need to be written manually. Current docs are here: https://raw.githubusercontent.com/envoyproxy/envoy/refs/heads/main/docs/root/configuration/http/http_filters/lua_filter.rst

zhishi commented on envoy/ssl/connection.h:
```diff
@@ -95,6 +96,11 @@ class ConnectionInfo {
+  /**
+   * @return the well-known attribute values parsed from subject field of the peer certificate.
+   **/
+  virtual const ParsedX509NameConstSharedPtr& parsedSubjectPeerCertificate() const PURE;
```
@ggreenway just want to confirm it's still ok to use shared_ptr in the member right (so we don't allocate memory until someone called the function)?

ggreenway replied:
Use a `unique_ptr` instead

zhishi replied:
@ggreenway  if return reference then we can't use null pointer and caller can't check if the subject not exist, such as the change in quic_ssl_connection_info.h which must allocate the object. so looks OptRef is the only choice. 

That being said, my understanding is that the OptRef is just a wrapper of raw pointer, which is less safe than shared_ptr because it can still keep the pointer when the original object is deleted and will be used to reference an invalid address. shared_ptr will make sure the object is kept, although it has risk of memory leak but still better than invalid memory access I guess.

zhishi commented on source/extensions/filters/common/lua/wrappers.cc:
```diff
@@ -296,6 +307,21 @@ int SslConnectionWrapper::luaSubjectPeerCertificate(lua_State* state) {
+int SslConnectionWrapper::luaParsedSubjectPeerCertificate(lua_State* state) {
+  const auto& parsed_name = connection_info_.parsedSubjectPeerCertificate();
+  if (parsed_name != nullptr) {
+    if (parsed_subject_peer_certificate_.get() != nullptr) {
```
I followed the similar code in the line 404 int ConnectionWrapper::luaSsl(lua_State* state), my understanding is that we may want to make sure return value always based the truth from  connection_info_ rathe than the cached parsed_subject_peer_certificate_ value.

zhishi commented:
@ggreenway I have addressed comment and updated document. Do you think we should also include other well-known fields this time or only add it when someone need it in future?

ggreenway commented on source/extensions/filters/common/lua/wrappers.h:
```diff
@@ -117,6 +117,33 @@ class MetadataMapWrapper : public BaseLuaObject<MetadataMapWrapper> {
+   */
+  DECLARE_LUA_FUNCTION(ParsedX509NameWrapper, luaOrganizationName);
+
+  const Ssl::ParsedX509Name& parsed_name_;
```
I wonder if it would make more sense to have this be a value instead of a reference. It's not a large struct, and then there's no concern about lifetimes, you don't need to mark-dead, etc.

zhishi replied:
I just followed the similar code in line 314 for connection_info_. currently I only added CN and O, but eventually it should contain more fields which will increase the size and cost. I assume envoy code base want optimize code as fast as possible.

ggreenway commented on source/extensions/filters/common/lua/wrappers.h:
```diff
@@ -274,7 +308,11 @@ class SslConnectionWrapper : public BaseLuaObject<SslConnectionWrapper> {
+  void onMarkDead() override { parsed_subject_peer_certificate_.reset(); }
+
   const Ssl::ConnectionInfo& connection_info_;
+  LuaDeathRef<ParsedX509NameWrapper> parsed_subject_peer_certificate_;
```
Why does this need to be LuaDeathRef, and `connection_info_` above is a bare reference?

zhishi replied:
because the connection_info_ is an internal object used by code, while the parsed_subject_peer_certificate_ is actually a cached lua object will be returned to user. you can see similar usage in line 340 for ConnectionWrapper.

ggreenway reviewed (changes requested):
/wait

zhishi commented:
@ggreenway I have addressed the comments. For some reason the CI job Envoy/Checks always fail on step ./ci/do_ci.sh coverage:
Run rm -rf "$TMP_STDOUT"
Run failed
Error: Process completed with exit code 1.

I feel it's not related with my change.

ggreenway commented:
> @ggreenway I have addressed the comments. For some reason the CI job Envoy/Checks always fail on step ./ci/do_ci.sh coverage: Run rm -rf "$TMP_STDOUT" Run failed Error: Process completed with exit code 1.
> 
> I feel it's not related with my change.

If you scroll up a little ways in the coverage log, you'll see

```
Code coverage 96.6 is good and higher than limit of 96.1
Checking per-extension coverage
Per-extension coverage failed:
Code coverage for source/common/tls is lower than limit of 95.3 (95.2)
```

I think this is due to your change. Please add more test coverage. 
/wait

ggreenway commented:
There's a merge conflict. Also a bunch of CI changes have been made recently. Please merge main into your PR branch and push.

/wait

ggreenway commented:
Also, I just re-worked how data is cached to counter-act the growing memory use per connection in https://github.com/envoyproxy/envoy/pull/36767. You'll need to integrate into that. But as I said earlier, I think it would make sense to return the unique_ptr by value, and NOT hold it in the ConnectionInfoImplBase at all.

zhishi commented:
oh, that's a good change, let me integrate it.

zhishi commented:
> Also, I just re-worked how data is cached to counter-act the growing memory use per connection in #36767. You'll need to integrate into that. But as I said earlier, I think it would make sense to return the unique_ptr by value, and NOT hold it in the ConnectionInfoImplBase at all.

do you mean don't want cache the ParsedX509Name object at all, just generate it every time when the function is called?

zhishi commented:
actually, based on your code change, when I return a unique_ptr, it will return a const unique_ptr& to caller, so it's still cached and won't be deleted by caller, is my understanding right?

ggreenway commented:
I'm debating whether it's useful to cache the object; but if non-lua accessors are added, it may be worthwhile. 

The internal `getCachedValueOrCreate` could return a `unique_ptr<>&`, but `parsedSubjectPeerCertificate()` should not.

zhishi commented:
I feel it won't hurt to cache the object, because even connectionInfo don't cache it, the lua wrapper will still cache it anyway. and in typical usage the lua filter written by user (my use case at least) will be in the route level, which will be called on every requests go through that connection, so this function will be called multiple times.

ggreenway commented:
> I feel it won't hurt to cache the object, because even connectionInfo don't cache it, the lua wrapper will still cache it anyway. and in typical usage the lua filter written by user (my use case at least) will be in the route level, which will be called on every requests go through that connection, so this function will be called multiple times.

Good point; I agree it makes sense to cache.

zhishi commented:
@ggreenway I have updated the code with your change, and I think the code is more cleaner with your cache interface. since this change several places in my code, may need your review all the changes again. thanks.

ggreenway commented on envoy/ssl/connection.h:
```diff
@@ -95,6 +96,12 @@ class ConnectionInfo {
+   * @return the well-known attribute values parsed from subject field of the peer certificate.
+   *         Returns absl::nullopt if there is no peer certificate.
+   **/
+  virtual const ParsedX509NamePtr& parsedSubjectPeerCertificate() const PURE;
```
Don't return a reference to a unique_ptr; that's a weird interface. Switch back to the OptRef you had.

zhishi replied:
I actually like the reference to unique_ptr than the OptRef interface. 
1. the usage is no much difference from caller, just use as a pointer for both cases. 
2. using the reference to unique_ptr actually make code cleaner. 
3. since we named the unique_ptr to ParsedX509NamePtr now all the return type of interface in connectionInfo is const XXX& which is more consistent and less confusing
4. it has less chance to let caller hold an invalid pointer like the OptRef because a const unique_ptr is not allowed to copy/move in general.

ggreenway replied:
It's not the correct way to treat unique_ptr in c++. Please fix.

ggreenway commented on source/common/tls/connection_info_impl_base.cc:
```diff
@@ -380,6 +380,17 @@ const std::string& ConnectionInfoImplBase::subjectPeerCertificate() const {
 }
 
+const Ssl::ParsedX509NamePtr& ConnectionInfoImplBase::parsedSubjectPeerCertificate() const {
+  return getCachedValueOrCreate<Ssl::ParsedX509NamePtr>(
```
Right here is where you can convert from unique_ptr to OptRef depending on if it's nullptr or not.

ggreenway reviewed (changes requested):
/wait

ggreenway reviewed (approved):
LGTM

ggreenway reviewed (changes requested):
Sorry, one last thing: please add an entry in changelog/current.yaml

/wait
